File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def hasNode(self, root, node) -> bool:
        if root is None:
            return False
        if root.val == node.val:
            return True
        return self.hasNode(root.left, node) or self.hasNode(root.right, node)
        
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        leftValid = self.hasNode(root.left, p) and self.hasNode(root.left, q)
        rightValid = self.hasNode(root.right, p) and self.hasNode(root.right, q)
        if not leftValid and not rightValid:
            return root
        if leftValid:
            return self.lowestCommonAncestor(root.left, p, q)
        else:
            return self.lowestCommonAncestor(root.right, p, q)
        

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        LinkedList<TreeNode> q = new LinkedList<>();
        int maxDepth = 0;
        int width = 1;
        q.add(root);
        while (!q.isEmpty()) {
            maxDepth++;
            for (int i = 0; i < width; i++) {
                TreeNode node = q.removeFirst();
                if (node.left != null) {
                    q.add(node.left);
                }
                if (node.right != null) {
                    q.add(node.right);
                }
            }
            width = q.size();
        }
        return maxDepth;
    }
}
